Mathlib/Tactic/Order.lean

```diff
@@ -0,0 +1,491 @@
+    MetaM Graph := do
+  let nltFacts := facts.filter fun fact => match fact.rel with | .nlt => true | _ => false
+  let mut used : Array Bool := mkArray nltFacts.size false
+  let mut g := g
+  while true do
+    let mut changed : Bool := false
+    for i in [:nltFacts.size] do
+      if used[i]! then
+        continue
+      let fact := nltFacts[i]!
+      let .some pf ← buildTransitiveLeProof g fact.lhs fact.rhs (idxToAtom.get! fact.rhs) | continue
+      changed := true
+      used := used.set! i true
+      let newFact : AtomicFact := ⟨fact.rhs, fact.lhs, .le,
+        ← mkAppM ``le_of_not_lt_le #[fact.proof, pf]⟩
+      g := g.modify fact.rhs fun edges => edges.push newFact
+    if !changed then
+      break
+  return g
+
+/-- Using the `≤`-graph `g`, find a contradiction with some `≰`-fact. -/
+def findContradictionWithNle (g : Graph) (idxToAtom : Std.HashMap ℕ Expr)
+    (facts : Array AtomicFact) : MetaM <| Option Expr := do
+  for fact in facts do
+    if fact.rel != .nle then
+      continue
+    let .some pf ← buildTransitiveLeProof g fact.lhs fact.rhs
+      (idxToAtom.get! fact.rhs) | continue
+    return .some <| mkApp fact.proof pf
+  return .none
+
+/-- Supported order types: linear, partial, and preorder. -/
+inductive OrderType
+| lin | part | pre
+deriving BEq
+
+/-- Find the "best" instance of an order on a given type. A linear order is preferred over a partial
+order, and a partial order is preferred over a preorder. -/
+def findBestInstance (type : Expr) : MetaM <| Option OrderType := do
+  if (← synthInstance? (← mkAppM ``LinearOrder #[type])).isSome then
+    return .some .lin
+  if (← synthInstance? (← mkAppM ``PartialOrder #[type])).isSome then
+    return .some .part
+  if (← synthInstance? (← mkAppM ``Preorder #[type])).isSome then
+    return .some .pre
+  return .none
+
+/-- A finishing tactic for solving goals in arbitrary `Preorder`, `PartialOrder`,
+or `LinearOrder`. -/
+elab "order" : tactic => focus do
+  let g ← getMainGoal
+  let .some g ← g.falseOrByContra | return
+  setGoals [g]
+  let TypeToAtoms ← collectFacts g
+  g.withContext do
+  for (type, (idxToAtom, facts)) in TypeToAtoms do
+    let .some orderType ← findBestInstance type | continue
+    let facts : Array AtomicFact ← match orderType with
+    | .lin => preprocessFactsLinear g facts
+    | .part => preprocessFactsPartial g facts
+    | .pre => preprocessFactsPreorder g facts
+    let mut graph := constructLeGraph idxToAtom.size facts
+    if orderType != .lin then
+      graph ← updateGraphWithNlt graph idxToAtom facts
+    if orderType == .pre then
+      let .some pf ← findContradictionWithNle graph idxToAtom facts | continue
+      g.assign pf
+      return
+    else
+      let .some contNe := findContradictoryNe graph facts | continue
+      let .some pf1 ← buildTransitiveLeProof graph contNe.lhs contNe.rhs
+        (idxToAtom.get! contNe.rhs) | throwError "bug"
```

Comment:
Can this give a more descriptive error? In the unlikely situation that an end user hits this or the below error message, it would be nice to know which one it is, and which algorithm assumption broke.